File: src/main/java/model/util/PositionIterator.java
package model.util;

import org.jetbrains.annotations.NotNull;

import java.util.Iterator;

/**
 * Iterator that loops through all the position on the board
 */
public class PositionIterator implements Iterator<Position> {
    private static final Position MAXIMUM = new Position(Position.LIMIT, 0);

    @NotNull
    private Position position = new Position(0, 0);

    @Override
    public boolean hasNext() {
        return !position.equals(MAXIMUM);
    }

    @NotNull
    @Override
    public Position next() {
        Position positionToReturn = position;

        if (position.getColumn() == Position.LIMIT - 1) {
            position = new Position(position.getRow() + 1, 0);
        } else {
            position = position.shift(Offset.RIGHT);
        }

        return positionToReturn;
    }
}

File: engine/src/main/java/engine/util/PositionIterator.kt
package engine.util

/**
 * Iterator that loops through all the position on the pieceMap
 */
class PositionIterator : Iterator<Position> {
    private val maximum = Position(LIMIT, 0) //One row below max

    private var nextPosition = Position(0, 0)

    override fun hasNext(): Boolean {
        return nextPosition != maximum
    }

    override fun next(): Position {
        val currentPosition = nextPosition

        nextPosition += SHIFT_RIGHT

        if (nextPosition.column == LIMIT) nextPosition = Position(nextPosition.row + 1, 0)

        return currentPosition
    }
}
